File: ncRegras/rVenda.vb
Imports System.Transactions

Imports ncDados.nsVenda
Imports ncPersistencia.nsVenda
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao

Namespace nsVenda

  Public Class rVenda

    '-- Métodos de controle ( Várias chamadas; Controle de transação )

    Public Function Listar() As ColecaoVenda

      Dim retorno As ColecaoVenda
      Dim persistencia As pVenda
      Dim retornoPersistencia As ColecaoVenda

      Try

        retorno = New ColecaoVenda

        persistencia = New pVenda
        retornoPersistencia = persistencia.Listar()

        If Not retornoPersistencia Is Nothing Then
          If retornoPersistencia.Count > 0 Then
            retorno.AddRange(retornoPersistencia)
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Listar Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Listar = retorno

    End Function

    Public Function Consultar(ByVal dados As dVenda) As ColecaoVenda

      Dim retorno As ColecaoVenda
      Dim persistencia As pVenda
      Dim retornoPersistencia As ColecaoVenda

      Try

        retorno = New ColecaoVenda

        persistencia = New pVenda
        retornoPersistencia = persistencia.Consultar(dados)

        If Not retornoPersistencia Is Nothing Then
          If retornoPersistencia.Count > 0 Then
            retorno.AddRange(retornoPersistencia)
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function
        Public Function ConsultarPix(ByVal dados As dVenda) As ColecaoVenda

            Dim retorno As ColecaoVenda
            Dim persistencia As pVenda
            Dim retornoPersistencia As ColecaoVenda

            Try

                retorno = New ColecaoVenda

                persistencia = New pVenda
                retornoPersistencia = persistencia.ConsultarPix(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Return retorno

        End Function
        Public Function ConsultarVendasPorVendedor(ByVal dados As dVendasPorVendedor) As ColecaoVendasPorVendedor

            Dim retorno As ColecaoVendasPorVendedor
            Dim persistencia As pVenda
            Dim retornoPersistencia As ColecaoVendasPorVendedor

            Try

                retorno = New ColecaoVendasPorVendedor

                persistencia = New pVenda
                retornoPersistencia = persistencia.ConsultarVendasPorVendedor(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Return retorno

        End Function
        Public Function ConsultarVendasDaLoja(ByVal dados As dVendasPorVendedor) As ColecaoVendasPorVendedor

            Dim retorno As ColecaoVendasPorVendedor
            Dim persistencia As pVenda
            Dim retornoPersistencia As ColecaoVendasPorVendedor

            Try

                retorno = New ColecaoVendasPorVendedor

                persistencia = New pVenda
                retornoPersistencia = persistencia.ConsultarVendasDaLoja(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Return retorno

        End Function
        Public Function ConsultarCrediarioPix(ByVal dados As dVenda) As ColecaoVenda

            Dim retorno As ColecaoVenda
            Dim persistencia As pVenda
            Dim retornoPersistencia As ColecaoVenda

            Try

                retorno = New ColecaoVenda

                persistencia = New pVenda
                retornoPersistencia = persistencia.ConsultarCrediarioPix(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Return retorno

        End Function
        Public Function ConsultarTroca(ByVal dados As dVenda) As ColecaoVenda

      Dim retorno As ColecaoVenda
      Dim persistencia As pVenda
      Dim retornoPersistencia As ColecaoVenda

      Try

        retorno = New ColecaoVenda

        persistencia = New pVenda
        retornoPersistencia = persistencia.ConsultarTroca(dados)

        If Not retornoPersistencia Is Nothing Then
          If retornoPersistencia.Count > 0 Then
            retorno.AddRange(retornoPersistencia)
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function

    Public Function ConsultarUltimaVenda(ByVal produto_cid As Integer) As dVenda

      Dim retorno As dVenda
      Dim persistencia As pVenda

      Try

        retorno = New dVenda

        persistencia = New pVenda
        retorno = persistencia.ConsultarUltimaVenda(produto_cid)

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em ConsultarUltimaVenda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      ConsultarUltimaVenda = retorno

    End Function

    Public Function ConsultarMax() As Integer

      Dim persistencia As pVenda
      Dim retornoPersistencia As Integer

      Try

        persistencia = New pVenda
        retornoPersistencia = persistencia.ConsultarMax()

      Catch ex As Exception

        retornoPersistencia = Nothing
        Throw New ExcecaoNascomercio("Erro em ConsultarMax Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      ConsultarMax = retornoPersistencia

    End Function

    Public Function Incluir(ByVal dados As dVenda, ByVal dadosProdutos As ColecaoVendaProduto) As Integer

      Dim retorno As Integer
      Dim persistencia As pVenda
      Dim persistenciaProduto As pVendaProduto

      Try

        Using ts As New TransactionScope

          persistencia = New pVenda
          retorno = persistencia.Incluir(dados)

          persistenciaProduto = New pVendaProduto()
          If Not IsNothing(dadosProdutos) Then
            For Each produto As dVendaProduto In dadosProdutos
              produto.controle = retorno
              persistenciaProduto.Incluir(produto)
            Next
          End If

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Incluir = retorno

    End Function

    Public Function IncluirTroca(ByVal dados As dVenda, ByVal dadosProdutos As ColecaoVendaProduto) As Integer

      Dim retorno As Integer
      Dim persistencia As pVenda
      Dim persistenciaProduto As pVendaProduto

      Try

        Using ts As New TransactionScope

          persistencia = New pVenda
          retorno = persistencia.IncluirVale(dados)

          persistenciaProduto = New pVendaProduto()
          If Not IsNothing(dadosProdutos) Then
            For Each produto As dVendaProduto In dadosProdutos
              produto.controle = retorno
              persistenciaProduto.IncluirTroca(produto)
            Next
          End If

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function

    Public Function IncluirVale(ByVal dados As dVenda) As Integer

      Dim retorno As Integer
      Dim persistencia As pVenda

      Try

        Using ts As New TransactionScope

          persistencia = New pVenda
          retorno = persistencia.IncluirVale(dados)

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Vale [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function

    Public Function IncluirCrediarioPagamento(ByVal dados As dVenda) As Integer

      Dim retorno As Integer
      Dim persistencia As pVenda

      Try

        Using ts As New TransactionScope

          persistencia = New pVenda
          retorno = persistencia.IncluirCrediarioPagamento(dados)

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Pagamento de Crediário [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function

    Public Function Alterar(ByVal dados As dVenda) As Integer

      Dim retorno As Integer
      Dim persistencia As pVenda

      Try

        Using ts As New TransactionScope
          persistencia = New pVenda
          retorno = persistencia.Alterar(dados)

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Alterar Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Alterar = retorno

    End Function

        Public Function Alterar(ByVal controle As String, ByVal chave As String) As Integer

            Dim retorno As Integer
            Dim persistencia As pVenda

            Try

                Using ts As New TransactionScope
                    persistencia = New pVenda
                    retorno = persistencia.Alterar(controle, chave)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Venda [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Alterar = retorno

        End Function
        Public Function Excluir(ByVal dados As dVenda) As Integer

      Dim retorno As Integer
      Dim persistencia As pVenda

      Try

        Using ts As New TransactionScope
          persistencia = New pVenda
          retorno = persistencia.Excluir(dados)
          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Excluir Venda [" & Me.ToString() & "] - " & ex.Message)

      End Try

            Return retorno

    End Function

        Public Function ExcluirVale(ByVal dados As dVenda) As Integer

            Dim retorno As Integer
            Dim persistencia As pVenda

            Try

                Using ts As New TransactionScope
                    persistencia = New pVenda
                    retorno = persistencia.ExcluirVale(dados)
                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Vale [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Return retorno

        End Function

        Public Function ConsultarFechamento(ByVal dados As dVenda) As ColecaoVenda

            Dim retorno As ColecaoVenda
            Dim persistencia As pVenda
            Dim retornoPersistencia As ColecaoVenda

            Try

                retorno = New ColecaoVenda

                persistencia = New pVenda
                retornoPersistencia = persistencia.ConsultarFechamento(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message)

            End Try

            Return retorno

        End Function
    End Class

End Namespace

File: Servicos/rVenda.cs
using System;
using System.Transactions;
using Comum;
using Repositorios;
using Modelos;

namespace Servicos
{
    public class rVenda : IsVenda
    {
        private readonly IpVenda _venda;
        private readonly IpVendaProduto _vendaProduto;

        public rVenda(IpVenda venda, IpVendaProduto vendaProduto)
        {
            _venda = venda;
            _vendaProduto = vendaProduto;
        }

        public ColecaoVenda Listar()
        {
            try { return _venda.Listar(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Listar Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaodVendasNfe ListarVendasNfe(string dataIni, string dataFim)
        {
            try { return _venda.ListarVendasNfe(dataIni, dataFim); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ListarVendasNfe Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaodVendasABC ListarVendasABC(string dataIni, string dataFim, string tipo)
        {
            try { return _venda.ListarVendasABC(dataIni, dataFim, tipo); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ListarVendasABC Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVenda Consultar(dVenda dados)
        {
            try { return _venda.Consultar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Consultar Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVenda ConsultarPix(dVenda dados)
        {
            try { return _venda.ConsultarPix(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarPix Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVendasPorVendedor ConsultarVendasPorVendedor(dVendasPorVendedor dados)
        {
            try { return _venda.ConsultarVendasPorVendedor(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarVendasPorVendedor Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVendasPorVendedor ConsultarVendasDaLoja(dVendasPorVendedor dados)
        {
            try { return _venda.ConsultarVendasDaLoja(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarVendasDaLoja Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVenda ConsultarCrediarioPix(dVenda dados)
        {
            try { return _venda.ConsultarCrediarioPix(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarCrediarioPix Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVenda ConsultarTroca(dVenda dados)
        {
            try { return _venda.ConsultarTroca(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarTroca Venda [" + ToString() + "] - " + ex.Message); }
        }

        public dVenda ConsultarUltimaVenda(int produto_cid)
        {
            try { return _venda.ConsultarUltimaVenda(produto_cid); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarUltimaVenda Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int ConsultarMax()
        {
            try { return _venda.ConsultarMax(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarMax Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int Incluir(dVenda dados, ColecaoVendaProduto dadosProdutos)
        {
            try
            {
                using (var ts = new TransactionScope())
                {
                    int retorno = _venda.Incluir(dados);
                    if (dadosProdutos != null)
                        foreach (var produto in dadosProdutos)
                        {
                            produto.controle = retorno;
                            _vendaProduto.Incluir(produto);
                        }
                    ts.Complete();
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Incluir Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int IncluirnNF(dBasennf dados)
        {
            try { return _venda.IncluirnNF(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em IncluirnNF Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int IncluirTroca(dVenda dados, ColecaoVendaProduto dadosProdutos)
        {
            try
            {
                using (var ts = new TransactionScope())
                {
                    int retorno = _venda.IncluirVale(dados);
                    if (dadosProdutos != null)
                        foreach (var produto in dadosProdutos)
                        {
                            produto.controle = retorno;
                            _vendaProduto.IncluirTroca(produto);
                        }
                    ts.Complete();
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em IncluirTroca Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int IncluirVale(dVenda dados)
        {
            try { return _venda.IncluirVale(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em IncluirVale Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int IncluirCrediarioPagamento(dVenda dados)
        {
            try { return _venda.IncluirCrediarioPagamento(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em IncluirCrediarioPagamento Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int Alterar(dVenda dados)
        {
            try { return _venda.Alterar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Alterar Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int Alterar(string controle, string chave)
        {
            try { return _venda.Alterar(controle, chave); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Alterar Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int AlterarBaseNnf(dBasennf dados)
        {
            try { return _venda.AlterarBaseNnf(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em AlterarBaseNnf Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int Excluir(dVenda dados)
        {
            try { return _venda.Excluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Excluir Venda [" + ToString() + "] - " + ex.Message); }
        }

        public int ExcluirVale(dVenda dados)
        {
            try { return _venda.ExcluirVale(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ExcluirVale Venda [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVenda ConsultarFechamento(dVenda dados)
        {
            try { return _venda.ConsultarFechamento(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarFechamento Venda [" + ToString() + "] - " + ex.Message); }
        }
    }
}

